lib/mongoose_push_web/plugs/cast_and_validate.ex

```diff
@@ -0,0 +1,133 @@
+defmodule MongoosePushWeb.Plug.CastAndValidate do
+  @behaviour Plug
+
+  @impl Plug
+  def init(opts) do
+    OpenApiSpex.Plug.CastAndValidate.init(opts)
+  end
+
+  @impl Plug
+  def call(conn = %{params: %{"service" => _, "alert" => _}}, opts = %{operation_id: _}) do
+    OpenApiSpex.Plug.CastAndValidate.call(conn, opts)
+  end
+
+  @impl Plug
+  def call(conn = %{params: %{"service" => _, "data" => _}}, opts = %{operation_id: _}) do
+    OpenApiSpex.Plug.CastAndValidate.call(conn, opts)
+  end
+
+  @impl Plug
+  def call(conn = %{params: %{"service" => _}}, opts = %{operation_id: _}) do
+    update_schema_and_do_call(conn, opts)
+  end
+
+  @impl Plug
+  def call(conn = %{params: %{"alert" => _}}, opts = %{operation_id: _}) do
+    update_schema_and_do_call(conn, opts)
+  end
+
+  @impl Plug
+  def call(conn = %{params: %{"data" => _}}, opts = %{operation_id: _}) do
+    update_schema_and_do_call(conn, opts)
+  end
+
+  @impl Plug
+  def call(
+        conn = %{
+          private: %{
+            phoenix_controller: controller,
+            phoenix_action: action,
+            open_api_spex: private_data
+          }
+        },
+        opts
+      ) do
+    operation =
+      case private_data.operation_lookup[{controller, action}] do
+        nil ->
+          operationId = controller.open_api_operation(action).operationId
+          operation = private_data.operation_lookup[operationId]
+
+          operation_lookup =
+            private_data.operation_lookup
+            |> Map.put({controller, action}, operation)
+
+          OpenApiSpex.Plug.Cache.adapter().put(
+            private_data.spec_module,
+            {private_data.spec, operation_lookup}
+          )
+
+          operation
+
+        operation ->
+          operation
+      end
+
+    if operation.operationId do
+      call(conn, Map.put(opts, :operation_id, operation.operationId))
+    else
+      raise "operationId was not found in action API spec"
+    end
+  end
+
+  @impl Plug
+  def call(conn, opts) do
+    OpenApiSpex.Plug.CastAndValidate.call(conn, opts)
+  end
+
+  # When OpenApiSpex matches incoming request to schema defined as `oneOf: [%Reference{}]`, 
```

Comment:
This can be a `@doc` or `@moduledoc` whichever you find more appropriate.

Reply:
From the hexdoc you linked above:

> Elixir treats documentation and code comments as different concepts. Documentation is an explicit contract between you and users of your Application Programming Interface (API), be them third-party developers, co-workers, or your future self. Modules and functions must always be documented if they are part of your API.

> **Code comments** are aimed at developers reading the code. They are useful for marking improvements, **leaving notes (for example, why you had to resort to a workaround due to a bug in a library)**, and so forth. They are tied to the source code: you can completely rewrite a function and remove all existing code comments, and it will continue to behave the same, with no change to either its behaviour or its documentation.